components/script/dom/canvasrenderingcontext.rs

```diff
@@ -0,0 +1,1401 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasFillRule;
+use crate::dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasImageSource;
+use crate::dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasLineCap;
+use crate::dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasLineJoin;
+use crate::dom::bindings::codegen::Bindings::ImageDataBinding::ImageDataMethods;
+use crate::dom::bindings::codegen::UnionTypes::StringOrCanvasGradientOrCanvasPattern;
+use crate::dom::bindings::error::{Error, ErrorResult, Fallible};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::num::Finite;
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::canvasgradient::{CanvasGradient, CanvasGradientStyle, ToFillOrStrokeStyle};
+use crate::dom::canvaspattern::CanvasPattern;
+use crate::dom::canvasrenderingcontext2d::*;
+use crate::dom::element::cors_setting_for_element;
+use crate::dom::element::Element;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::htmlcanvaselement::{CanvasContext, HTMLCanvasElement};
+use crate::dom::imagedata::ImageData;
+use crate::dom::node::{Node, NodeDamage};
+use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+use crate::dom::textmetrics::TextMetrics;
+use crate::unpremultiplytable::UNPREMULTIPLY_TABLE;
+use canvas_traits::canvas::{Canvas2dMsg, CanvasId, CanvasMsg};
+use canvas_traits::canvas::{CompositionOrBlending, FillOrStrokeStyle, FillRule};
+use canvas_traits::canvas::{LineCapStyle, LineJoinStyle, LinearGradientStyle};
+use canvas_traits::canvas::{RadialGradientStyle, RepetitionStyle};
+use cssparser::Color as CSSColor;
+use cssparser::{Parser, ParserInput, RGBA};
+use euclid::{
+    default::{Point2D, Rect, Size2D, Transform2D},
+    vec2,
+};
+use ipc_channel::ipc::{self, IpcSender};
+use net_traits::image_cache::CanRequestImages;
+use net_traits::image_cache::ImageCache;
+use net_traits::image_cache::ImageOrMetadataAvailable;
+use net_traits::image_cache::ImageResponse;
+use net_traits::image_cache::ImageState;
+use net_traits::image_cache::UsePlaceholder;
+use net_traits::request::CorsSettings;
+use pixels::PixelFormat;
+use profile_traits::ipc as profiled_ipc;
+use script_traits::ScriptMsg;
+use servo_url::{ImmutableOrigin, ServoUrl};
+use std::cell::Cell;
+use std::str::FromStr;
+use std::sync::Arc;
+#[must_root]
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct CanvasState {
+    #[ignore_malloc_size_of = "Defined in ipc-channel"]
+    pub ipc_renderer: IpcSender<CanvasMsg>,
```

Comment:
Rather than making all of these members public, can we add getter methods for anything that needs to be accessed from canvasrenderingcontext2d.rs and offscreenrenderingcontext2d.rs?